성적관리 [STEP 1] wily.kim

wilybear opened the pull request:
@wonhee009 

성적관리 첫번째 스탭 PR입니다!

### 전체적인 구조
- GradeConsole : 콘솔창의 출력을 담당하는 구조체, 입력에 따라 메뉴를 보여줍니다
- GradeManager: student 데이터를 관리하는 클래스, 학생을 추가하거나 삭제하는 로직을 수행합니다.
위와 같은 구조를 한것은 역할에 따라 나누어봤습니다. 콘솔에 출력만을 담당하는 GradeConsole은 구조체로 만들었고 데이터 관련 로직을 처리하는 GradeManager는 클래스로 만들었습니다. 

### GradeConsole
콘솔의 각 입력은 요규사항의 동작조건을 만족시키기 위해 영어와 숫자안지 확인합니다.

MenuAction Enum로 각 메뉴 항목들을 나타내었습니다.
showMenu에서 입력을 받으면 enum과 switch를 사용하여 옵션에 맞는 메뉴를 보여줍니다.

GradeManager의 인스턴스를 가지고 있고 학생 추가와 삭제 시 받은 입력값을 GradeManger로 넘겨서 추가 및 삭제를 합니다.

### GradeManager
students를 Set으로 하여 효과적으로 학생이름이 중복되지 않게 하였습니다.
학생 추가,삭제(addStudent, deleteStudent) 메소드는 bool을 반환하여 성공과 실패를 나타내는데 데이터 자체를 검증하지 않고 set collection api를 호출하고 그 결과값으로 성공실패를 판단하였습니다.

### 고민 및 질문
- **MenuAction enumeration을 올바르게 사용한 것인지, 또 case naming이 올바른 것인지 궁금합니다.** MenuAction의 case들의 이름을 action의 형태로 나타내려다가(ex) addStudent) 함수 이름과 혼동될 수 있을 것 같아 addingStudent로 했는데 더 좋은 방법이 있는지 궁금합니다.
`case addStudent`
`case addingStudent`

- **GradeManager의 메소드와 파라미어의 convention 관련 고민입니다.**
GradeManger에서 학생을 추가하는 메소드를 보면 
`addStudent(by name: String)`
`addStudent(_ name: String)`
두개에서 고민을 했었는데 아래의 경우 addStudentName이라고 읽혀서 위를 선택했습니다. 어떤 방식이 더 선호되나요??

- GradeConsole에 show메소드들을 살펴보면 모두 유사한 로직을 수행합니다.
1. 출력 (ex "추가할 학생의 이름을 입력해주세요")
2. guard문을 통합 입력 유효성 검증
3. 받은 입력에 대한 결과 처리
**앞에 1,2 번 과정은 따로 메소드를 분리하여 별도의 메소드로 만드는 것이 좋을까요?** 만약 별도의 메소드로 만든다면 입력값이 invalid라면 error를 던지고 이를 받아서 에러처리를 하게 되면 코드양이 비슷할 것 같아서 현재는 따로 분리하지 않았습니다. 좋은 방법이 있을까요?

wilybear committed:
feat: Step1 메뉴 출력 및 잘못된 입력 처리, 학생 추가 삭제 기능 구현

메뉴 구현
학생 추가, 삭제 구현
모든 입력은 동작조건에 맞게 숫자, 영어인지 검증

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+}
+
+struct GradeConsole {
+    let gradeManager = GradeManager()
```
> GradeConsole : 콘솔창의 출력을 담당하는 구조체, 입력에 따라 메뉴를 보여줍니다

위의 설명으로 GradeConsole은 입/출력의 역할만 갖는다고 이해했는데요.
입/출력의 역할만 갖는 곳에서 성적을 관리하는 `GradeManager`를 가지고 있는 모습이 조금 어색한거 같습니다.
콘솔과 매니저의 역할을 분리해보는건 어떨까요?😀

wilybear replied:
GradeConsole에서 입력을 받아서 해당 입력을 처리하기위해서 GradeManager를 넣어줬습니다. 
외부에서 GradeManager를 생성해서 GradeConsole에 넣어주는 방식을 생각했는데 결국엔 GradeConsole이 GradeManager를 가지게 되는거 같아서 이걸 분리한다고 하니 어떻게 하는게 좋은 방법일지 생각이 나지 않습니다. 
아니면 GradeManager가 GradeConsole을 가지는게 맞을까요?

wonhee009 replied:
`GradeConsole`이 입/출력만 담당한다면 제가 생각했을 때 필요한 기능은
- 사용자 input 받기
- 때에 따라 필요한 텍스트 출력하기
- input validation
- validate된 값 반환

이 정도로 생각했는데요.
`GradeManager`는 `GradeConsole`이 반환한 validate를 받아 메뉴들을 실행하는게 역할이라고 생각했구요.

GradeManager와 GradeConsole이 서로 갖고 있는 모습보다는 서로 분리된게 좀 더 역할에 맞지 않나?라고 생각했습니다.
분리하는 방식이라면 main에서 콘솔과 매니저를 가지고 있는 모습도 괜찮지 않을까 생각이 듭니다.😀

wilybear replied:
답변 감사합니다!!!. 제가 역할에 따라 분리하는게 익숙하지 않아서 어렵네요 ㅠㅠ
그렇다면 `GradeManager`와 `GradeConsole`을 `main`이 가지고 있고 
 main에서 실행 흐름을 담당하며 `GradeManger`와 `GradeConsole`의 함수들을 호출해서 사용 라고 제가 이해한게 맞을까요?
 
 지금 아래에 로직처럼 수정 중인데 완전히 분리하려고 하니 서로 필요한 데이터가 있을때 마다 넘겨주어야하는 건지 고민이됩니다.

main.swift
GameManager와 GameConsole 초기화 -> GameManager.running 체크 -> main에서 GameConsole의 메뉴함수 호출
-> main에서 GameConsole의 입력값 받는 함수 호출 -> main에서 GameConsole의 입력값에 해당하는 메뉴 호출 -> main에서 다시 입력값을 받음 -> main에서 GameManager로 데이터 처리를 넘김 -> main에서 결과값을 GameConsole로 넘겨서 결과 출력

이런식으로 진행하는데 계속 막혀서 제가 잘못이해한거 같아 질문합니다 ㅠ

wilybear replied:
우선 `GameManager`를 위주로 실행흐름을 바꿔보았습니다. 이렇게 하니 `GradeConsole`은 그대로 입출력과 검증만 담당하게 되었고 다만 `GradeManager`와의 의존성은 남아있는 거 같습니다. 
한번 살펴봐주실수 있으신가요?

wonhee009 replied:
확인했습니다~ 입/출력 역할이 매니저를 갖는 부분이 어색했었는데 매니저가 입/출력 역할 객체도 매니징하도록 흐름이 변경되었네요.
이전보다 역할 분리가 잘되어 있는거 같습니다!👍
둘의 의존성을 완전히 떼는 것도 하나의 방법이지만 매니저가 입/출력 역할 객체를 가지고 있는 것도 괜찮은거 같습니다.
(입/출력이 매니저를 가지고 있기에는 주객이 전도된 느낌였어서 드렸던 의견였거든요☺️)

wilybear replied:
감사합니다! 저도 고민하다보니 매니저가 매니저의 역할이 아니라 저장소의 역할을 했던거 같아서 수정해보았습니다. 조언 감사합니다! ☺️

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+        case addingGrade = "3"
+        case deletingGrade = "4"
+        case averageGrade = "5"
+        case exit = "x"
```
`enum`과 `CaseIterable`을 사용하셨네요!👍
`enum MenuAction`은 사용자가 성적관리기에서 할 수 있는 메뉴들을 정의한거로 보여요.
이 안에서 사용자의 input과 비교하기 위한 String rawValue들이 정의되어 있고, 출력을 위한 description이 선언되어 있네요.
`MenuAction`이라는 이름을 보면 뭔가의 행위를 할것만 같은데 **비교 혹은 출력**을 위한 부분들만 정의되어져 있는거 같아요.
`enum Menu`로만해도 해당 의도를 잘 표현할 수 있을거 같은데 어떻게 생각하실까요?🤔

wonhee009 replied:
#### Q. MenuAction enumeration을 올바르게 사용한 것인지, 또 case naming이 올바른 것인지 궁금합니다.

- `CaseIterable`을 채택해서 `allCases`를 통해 사용자에게 원하는 기능을 출력한 점 정말 좋습니다👍
---
- [CaseIterable 공식문서](https://developer.apple.com/documentation/swift/caseiterable)에 따르면 **The synthesized allCases collection provides the cases in order of their declaration.** 즉, allCases 반환 array의 순서는 enum에 정의된 케이스 순서와 동일하다고 나와 있는데요. (별도의 sort 없이 wilybear의 코드도 제대로 동작했던 거구요.) 
- 그렇다면 이후에 학생을 추가, 삭제하는 메뉴의 번호가 서로 바뀐다면 어떻게 될까요? **2: 학생추가, 1: 학생삭제, 2: 성적추가(변경), 4: 성적삭제, 5: 평점보기, x: 종료** 이런식으로 출력되지 않을까요? 이후에 메뉴 번호가 변경되더라도 오름차순으로 출력될 수 있도록 수정해 보는건 어떨까요?😀
---
- 이전 코멘트에 남겼듯이 `MenuAction`이라면 `adding-`이 좀 더 맞는 네이밍인거 같아요.(`-Action`이 뭔가 행위를 의미하는거 같아서요.)
- `Menu`라면 `add-`가 좀 더 맞는 네이밍인거 같구요.(행위보다는 메뉴자체를 의미하는거 같아서요.) 이 부분에 대해서는 같이 얘기해봐도 좋을거 같아요.😀

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+    private func printMenu() {
+        print("원하는 기능을 입력해주세요")
+        let menu = MenuAction.allCases.map{"\($0.rawValue): \($0.description)"}
+            .joined(separator: ", ")
```
👍

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+struct GradeConsole {
+    let gradeManager = GradeManager()
+        
+    func showMenu() -> Bool {
+        printMenu()
+        guard let input = readLine(), validate(input) else {
+            print("뭔가 입력이 잘못되었습니다. 1~5 사이의 숫자 혹은 X를 입력해주세요")
+            return true
+        }
+        switch MenuAction(rawValue: input.lowercased()){
+        case .addingStudent:
+            showAddStudentMenu()
+        case .deletingStudent:
+            showDeleteStudentMenu()
+        case .addingGrade:
+            break
+        case .deletingGrade:
+            break
+        case .averageGrade:
+            break
+        case .exit:
+            print("프로그램을 종료합니다...")
+            return false
+        case .none:
+            print("뭔가 입력이 잘못되었습니다. 1~5 사이의 숫자 혹은 X를 입력해주세요")
+        }
+        return true
+    }
```
`showMenu`라는 메서드명을 보면 메뉴만 보여주는 메서드일거 같은데 실제로는 더 많은 일을 수행하고 있는거 같아요.
(메뉴 출력 -> input validation -> menu matching -> 각 메뉴에 맞는 텍스트 출력 -> 각 메뉴에 맞는 행위)
`showMenu`에서는 메뉴를 출력해주고 다른 일들도 좀 더 작게 나눠보는건 어떨까요?
(하나의 메서드가 하나의 일만 하도록, 메서드명을 보고 어떤 일을 하는지 알 수 있도록 하는걸 목표로 하면서요!)

그리고 `showMenu` 메서드의 반환값이 Bool 타입인 부분도 어색한거 같아요.
프로그램 종료인지 아닌지 알기 위해 Bool을 반환하도록 작성하신거 같아요.
이 부분도 프로그램 종료인지 아닌지를 확인할 수 있는 메서드를 하나 만들어서 거기서 확인하도록 하는 방법도 좋을거 같습니다.😀

wonhee009 replied:
#### Q. GradeConsole에 show메소드들을 살펴보면 모두 유사한 로직을 수행합니다.

- 해당 질문에 대해서는 `show-` 메서드들을 좀 더 역할에 따라 잘게 쪼갠 뒤에 다시 얘기해보면 좋을거 같습니다!😃

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+        return true
+    }
+    
+    private func showAddStudentMenu() {
+        print("추가할 학생의 이름을 입력해주세요")
+        guard let name = readLine(), validate(name) else {
+            print("입력이 잘못되었습니다. 다시 확인해주세요.")
+            return
+        }
+        switch gradeManager.addStudent(by: name) {
+        case true:
+            print("\(name) 힉셍을 추가했습니다.")
+        case false:
+            print("\(name)은 이미 존재하는 학생입니다. 추가하지 않습니다.")
+        }
+    }
+    
+    private func showDeleteStudentMenu() {
+        print("삭제할 학생의 이름을 입력해주세요")
+        guard let name = readLine(),validate(name) else {
+            print("입력이 잘못되었습니다. 다시 확인해주세요.")
+            return
+        }
+        switch gradeManager.deleteStudent(by: name) {
+        case true:
+            print("\(name) 학생을 삭제하였습니다.")
+        case false:
+            print("\(name) 학생을 찾지 못했습니다.")
+        }
+    }
```
`show-`라는 메서드인데 안에서 데이터를 바꾸고하는 부분들이 어색한거 같아요.
또한 `-Menu`로 네이밍해서 메뉴에 관한 출력만 할거 같은데 데이터를 바꾸고 성공 여부에 따라 결과를 출력해주는 부분도 어색한거 같구요.
이런 부분들도 메서드를 잘게 쪼개보는건 어떨까요?😀

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+class GradeManager {
+    var students = Set<String>()
+    
+    func addStudent(by name: String) -> Bool{
```
```suggestion
    func addStudent(by name: String) -> Bool {
```

wonhee009 replied:
#### Q. GradeManager의 메소드와 파라미어의 convention 관련 고민입니다.

- 적절하게 잘 사용해주신거 같습니다. `_`을 사용해서 파라미터명을 생략하는 경우는 파라미터에 뭐가 들어갈지 너무 명백한 경우 아니고서는 잘 사용하지 않게 되더라구요. (아무래도 팀원들도 코드를 쉽게 알아볼 수 있어야하니까요.) 현재 코드에서 `by`는 적절하다고 생각됩니다.👍

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+        return success
+    }
+    
+    func deleteStudent(by name: String) -> Bool{
```
```suggestion
    func deleteStudent(by name: String) -> Bool {
```

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+while(gradeConsole.showMenu()){}
+
+class GradeManager {
+    var students = Set<String>()
```
스텝1에서는 학생 추가, 삭제만 구현하긴 하지만 이후 스텝에서는 한명의 학생에게 다수의 (과목명, 성적)이 추가 될텐데요.
`Set<String>`으로만 관리하기에는 힘들지 않을까 생각이 드는데요.
혹시 이후 스텝에 대해 생각해두신 방향이 있을까요?🤔

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
 // 
 
 import Foundation
+
```
```suggestion
```

wonhee009 reviewed (changes requested):
@wilybear 첫 번째 프로젝트, 첫 번째 스텝, 첫 번째 PR 축하드립니다!🎉

`GradeConsole`, `GradeManager`를 정의해서 역할을 분리하려고 노력하신 부분 좋았습니다👍
다만 해당 부분들을 따로 파일로도 분리해보는건 어떨까요?👀
현재는 한 파일에 뭉쳐져 있다보니 이 기능은 여기에 있는게 맞나?하면서 헷갈릴 수 있을거 같아요.

남겨주신 질문에 대한 대답과 코드 관련해서 코멘트 남겨 두었으니 확인해주시고 잘 모르겠거나 의논하고 싶은 부분이 있다면 언제든 DM 보내주시면 감사하겠습니다☺️

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,120 @@
+    
+    private func validate(_ input: String) -> Bool{
+        let input = input.replacingOccurrences(of: " ", with: "")
+        return (!input.isEmpty && Double(input) != nil) || input.range(of: "[^a-zA-Z0-9]", options: .regularExpression ) == nil
```
`return`에 너무 많은 수식을 넣게되면 가독성이 떨어질 수 있습니다.🥲

wilybear committed:
feat: Step 1 구조 수정, Student 구조체 정의

main.swift에 있던 클래스와 구조체들을 별도의 팔일로 분리
Student 구조체 정의

wilybear committed:
feat: STEP1 GradeManager로 실행흐름 변경, 구조 변경

GradeConsole -> GradeManager로 실행 흐름 변경
GradeConsole은 이제 입출력만 담당하는 구조체로 변경
GradeConsole의 vadlidate 함수 리턴값 읽기 쉽게 수정

wilybear committed:
feat: STEP1 unit test 추가

GradeMangerTest: 학생 추가 삭제 시 경우 테스트
GradeConsoleTest: 각종 입력에 대한 검증 validate 테스트

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -0,0 +1,93 @@
+        print(menu)
+    }
+    
+    func showSelected(_ menu: Menu){
```
```suggestion
    func showSelected(_ menu: Menu) {
```

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+    
+    func validate(_ input: String) -> Bool{
```
```suggestion
    func validate(_ input: String) -> Bool {
```

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+    var running = true
+    let gradeConsole = GradeConsole()
+    
+    func step(){
```
```suggestion
    func step() {
```

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+    
+    private func addStudentProcess(){
```
```suggestion
    private func addStudentProcess() {
```

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+        gradeConsole.showAddStudentResult(addStudent(by: name), name: name)
+    }
+    
+    private func deleteStudentProcess(){
```
```suggestion
    private func deleteStudentProcess() {
```

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+    
+    @discardableResult
+    func addStudent(by name: String) -> Bool{
```
```suggestion
    func addStudent(by name: String) -> Bool {
```

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+    
+    @discardableResult
+    func deleteStudent(by name: String) -> Bool{
```
```suggestion
    func deleteStudent(by name: String) -> Bool {
```

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -0,0 +1,93 @@
+    }
+    
+    func showAddStudentResult(_ success: Bool, name: String) {
+        switch success {
+        case true:
+            print("\(name) 힉셍을 추가했습니다.")
+        case false:
+            print("\(name)은 이미 존재하는 학생입니다. 추가하지 않습니다.")
+        }
```
`switch`문으로 Bool 값을 판단하는 이유가 있을까요?

wilybear replied:
과제 조건 중에 else를 사용하지 말라는 조건이 있어서 `if success else` 대신 사용 했습니다. 혹시 
`print(success ? "\(name) 힉셍을 추가했습니다." : "\(name)은 이미 존재하는 학생입니다. 추가하지 않습니다.")`
이건 가독성이 떨어질까요?? if문으로 하는게 좋을까요?

wonhee009 replied:
아하... 현재 else를 사용하지 않는게 조건이였죠...😅
`print(success ? "\(name) 힉셍을 추가했습니다." : "\(name)은 이미 존재하는 학생입니다. 추가하지 않습니다.")` 이런 방식도 좋다고 생각합니다.

```swift
success ? print(addStudentSuccessMessage) : print(addStudentFailMessage)
print(success ? addStudentSuccessMessage : addStudentFailMessage)
```
이렇게 텍스트를 변수로 관리하고 삼항 연산자 사용은 어떤가요?
삼항 연산자에 문자열이 그대로 들어가면 가독성이 떨어질거 같습니다.

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -0,0 +1,93 @@
+        return (isNumberic || isAlphabet) && !input.isEmpty
+    }
+       
+    enum InputError: Error {
+        case invalidInput
+    }
```
`Error`를 사용해보셨군요!☺️
이를 통해 커스텀 에러를 throw할 수 있게되었네요.
`var description`같은 프로퍼티를 정의해서 에러별로 출력할 텍스트를 정의해주는건 어떨까요?
([이런걸](https://developer.apple.com/documentation/foundation/localizederror) 사용하봐도 좋을거 같구요.)
지금은 do catch를 사용해서 어떤 Error가 나오든 하나의 에러 텍스트를 출력하고 있는데요.
이런 프로퍼티를 이용하면 다양한 커스텀 에러에 맞는 에러 텍스트를 출력할 수 있습니다☺️

wilybear replied:
이 부분에 대해서 고민을 해봤습니다.
현재 상황은 `GradeConsole` 안에 `readInput`, `getMenuInput`, `getInput` 세가지 함수가 있습니다. 
`readInput`에서 에러를 throw하고 `getMenuInput`과 `getInput`이 `readInput`을 호출하고 error handling을 진행합니다.

`readInput`은 아무런 파라미터를 받지 않고 기본적인 입력값인지 (문자, 숫자인지 비어있지는 않은지)체크만 하기 때문에 `menuInput`에서 일어난 에러인지 `getInput`에서 일어난 에러인지 알 수 없기 때문에 `invalidInput` 하나만 던질 수 있고 각각의 함수에서 에러 처리 시 상황의 맞는 텍스트를 출력합니다.

커스텀 에러 케이스를 몇가지 더 만들고 추가하려고 하자 위와 같은 구조에서는 수정이 어려웠습니다. 오랫동안 고민한 결과 아래 두가지 경우를 생각해 봤습니다.

1. `menuInput`, `getInput`에서 각각 상황의 알맞는 커스텀 에러를 던지고 `GradeManager`에서 에러를 핸들링한다. 에러를 정확하게 사용하는 것 같지만 이 경우에서는 print를 `GameManager`에서 하게 되므로 모든 입출력을 `GradeConsole`에서 담당한다는 원칙에 어긋나는거 같습니다
2. 단순히 error의 description을 프린트. 즉 현재 코드에서 큰 수정없이 각 catch의 print부분을 "받은 에러가 아닌" 해당 함수에 알맞는 에러의  설명을 출력한다. 이 경우는 error를 제대로 사용하지 않는 것 같습니다.

위에 두 경우 외에 다른 방법이 있는지 궁금합니다. `GradeConsole`에서 입력을 받는 함수들(`readInput`, `getMenuInput`, `getInput`)의 구조도 한번 살펴봐주시면 감사하겠습니다.🙏

wonhee009 replied:
`getInput`, `getMenuInput`이 외부에서 사용하기 위해 뚫려 있는 메서드이고 이 메서드들이 `readInput`을 호출하는거로 이해했습니다.☺️
생각해주신 2가지 방법 모두 살펴 보았는데요.
고민하시는 점 모두 공감합니다!
이 부분은 현재 코드 살리고 이후에 스텝 진행하면서 좀 더 생각해보시면 좋을거 같은데 어떠실까요?😀

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -0,0 +1,58 @@
+    
+    private func addStudentProcess(){
+        guard let name = gradeConsole.getInput() else { return }
+        gradeConsole.showAddStudentResult(addStudent(by: name), name: name)
```
```swift
let result = addStudent(by: name)
gradeConsole.showAddStudentResult(result, name: name)
```
이렇게 분리하는건 어떨까요?
뭉쳐 있으니 어떤걸 넘기자고 하는지가 명확하게 보이지 않는 느낌이에요🥲

wonhee009 reviewed (changes requested):
많은 수정이 있었네요👍
작은 개선점들 남겨두었습니다! 확인 후에 모르는 부분이나 같이 얘기하고 싶은 부분은 언제든 환영이에요👐
(수정된 코드 보면서 적용된 코멘트들은 resolve 처리했는데요. 작업하시면서 적용하신 부분은 아래 댓글로 남겨주셔도 좋고 이모지를 남겨주셔도 좋을거 같아요!)

wilybear committed:
feat: STEP1 Step1 error enum, 들여쓰기, 가독성 관련 수정

GradeManager 들여쓰기 수정 및 addStudent, deleteStudent 가독성 수정
GradeConsole 학생 추가 삭제 결과 출력 함수 수정
Input Error Handling 과정 수정
GameConsole에서 입력 받을 시 에러 발생 -> GameManager에게 던짐 -> GameManager는 GameConsole에 에러 출략 요청

wonhee009 reviewed (approved):
@wilybear 수고하셨습니다!🎉
다음 스텝에서 보도록 해요!
